src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,136 @@
+            + "Please double-check it. Index: ";
+    private static final String INVALID_ELEMENT_MESSAGE = "The provided element is not present "
+            + "in the list. Please double-check it.";
+    private Object[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            elementData = grow();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        rangeCheckForAdd(index);
+        if (size == elementData.length) {
+            elementData = grow();
+        }
+        System.arraycopy(elementData, index, elementData,
+                index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        rangeCheckForRemoveAndSetAndGet(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        rangeCheckForRemoveAndSetAndGet(index);
+        System.arraycopy(elementData, index, elementData, index + 1, size + 1);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        rangeCheckForRemoveAndSetAndGet(index);
+        final T resultValue = (T) elementData[index];
+        if (index < size - 1) {
+            System.arraycopy(elementData, index + 1, elementData, index, size - index - 1);
+        }
+        elementData[size - 1] = null;
+        size--;
+        return resultValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        hasElement(element);
+        int index = getIndex(element);
+        T resultValue = get(index);
+        remove(index);
+        return resultValue;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        return size == 0;
+    }
+
+    private Object[] grow() {
+        int newCapacity = (int) (elementData.length * CAPACITY_MULTIPLIER);
+        return elementData = Arrays.copyOf(elementData, newCapacity);
+    }
+
+    private void rangeCheckForAdd(int index) {
+        if (index > size || index < INDEX_LOWER_BOUND) {
+            throw new ArrayListIndexOutOfBoundsException(INVALID_INDEX_MESSAGE + index);
+        }
+    }
+
+    private void rangeCheckForRemoveAndSetAndGet(int index) {
+        if (index >= size || index < INDEX_LOWER_BOUND) {
+            throw new ArrayListIndexOutOfBoundsException(INVALID_INDEX_MESSAGE + index);
+        }
+    }
+
+    private void hasElement(T element) {
```

Comment:
methods with names, which starts with `has` should return boolean.

Reply:
I have renamed the method (as there was not much point to return boolean in this case)